datasource.go

```diff
@@ -0,0 +1,205 @@
+package main
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+
+	"golang.org/x/net/context"
+
+	"github.com/bitly/go-simplejson"
+	"github.com/grafana/grafana-plugin-model/go/datasource"
+	plugin "github.com/hashicorp/go-plugin"
+	"golang.org/x/net/context/ctxhttp"
+)
+
+var httpClient = &http.Client{}
+
+type ClickhouseDatasource struct {
+	plugin.NetRPCUnsupportedPlugin
+}
+
+func (t *ClickhouseDatasource) Query(ctx context.Context, req *datasource.DatasourceRequest) (r *datasource.DatasourceResponse, err error) {
+	// catch all panics and override err return value
+	defer func() {
+		if r := recover(); r != nil {
+			err = fmt.Errorf("clickhouse plugin panicked: %#w", r)
+		}
+	}()
+
+	refId := req.Queries[0].RefId
+	modelJson, err := simplejson.NewJson([]byte(req.Queries[0].ModelJson))
+	if err != nil {
+		return nil, err
+	}
+
+	query := modelJson.Get("rawQuery").MustString()
+	request, err := createRequest(req, query)
+	if err != nil {
+		return nil, err
+	}
+
+	response, err := ctxhttp.Do(ctx, httpClient, request)
+	if err != nil {
+		return nil, err
+	}
+
+	if response.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("invalid status code. status: %v", response.Status)
+	}
+
+	defer response.Body.Close()
+
+	body, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	return parseResponse(body, refId)
+}
+
+func createRequest(req *datasource.DatasourceRequest, query string) (*http.Request, error) {
+	body := ""
+	method := http.MethodGet
+	headers := http.Header{}
+	url, err := url.Parse(req.Datasource.Url)
+	if err != nil {
+		return nil, fmt.Errorf("unable to parse clickhouse url: %w", err)
+	}
+
+	params := url.Query()
+	params.Add("query", query+" FORMAT JSON")
+
+	// TODO Fix basic authorization. We have access to basicAuthPassword but not
+	// the basic auth name. Users will have to use the useYandexCloudAuthorization
+	// option instead for clickhouse auth.
+	secureOptions := req.Datasource.DecryptedSecureJsonData
+	options := make(map[string]interface{})
+	err = json.Unmarshal([]byte(req.Datasource.JsonData), &options)
+	if err != nil {
+		return nil, fmt.Errorf("unable to parse clickhouse options: %w", err)
+	}
+
+	for k, v := range options {
+		switch k {
+		case "usePOST":
+			method = http.MethodPost
+			params.Del("query")
+			body = query
+			break
```

Comment:
You shouldn't need to break within a go switch.

Reply:
good point. Is it better clarity to leave it in or leave it out?